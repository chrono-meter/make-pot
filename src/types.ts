import { pkgJsonHeaders, pluginHeaders, themeHeaders } from './extractors-maps'
import { GetTextTranslation } from 'gettext-parser'

export type ThemeHeadersType = keyof typeof themeHeaders
export type PluginHeadersType = keyof typeof pluginHeaders
export type PkgHeadersType = keyof typeof pkgJsonHeaders

/**
 * The args headers Object types
 */
export type PotHeaders =
	| (typeof pkgJsonHeaders)[PkgHeadersType]
	| (typeof pluginHeaders)[PluginHeadersType]
	| (typeof themeHeaders)[ThemeHeadersType]
	| 'fileComment'
	| 'packageName'

// type is the value of the themeHeader Object
export type DomainType =
	| 'plugin'
	| 'theme'
	| 'block'
	| 'theme-block'
	| 'generic'

/**
 * The patterns to use when extracting strings from files.
 *
 * @param {string} mergePaths - Comma-separated list of POT files whose contents should be merged with the extracted strings.
 *   If left empty, defaults to the destination POT file. POT file headers will be ignored.
 * @param {string} subtractPaths - Comma-separated list of POT files whose contents should act as some sort of denylist
 *   for string extraction. Any string which is found on that denylist will not be extracted. This can be useful when
 *   you want to create multiple POT files from the same source directory with slightly different content and no duplicate
 *   strings between them.
 * @param {boolean} subtractAndMerge - Whether source code references and comments from the generated POT file should be
 *   instead added to the POT file used for subtraction. Warning: this modifies the files passed to `subtractPaths`!
 * @param {string} include - Comma-separated list of files and paths that should be used for string extraction.
 *   If provided, only these files and folders will be taken into account for string extraction.
 *   For example, `--include="src,my-file.php` will ignore anything besides `my-file.php` and files in the `src`
 *   directory. Simple glob patterns can be used, i.e. `--include=foo-*.php` includes any PHP file with the `foo-`
 *   prefix. Leading and trailing slashes are ignored, i.e. `/my/directory/` is the same as `my/directory`.
 * @param {string} exclude - Comma-separated list of files and paths that should be skipped for string extraction.
 *   For example, `--exclude=.github,myfile.php` would ignore any strings found within `myfile.php` or the `.github`
 *   folder. Simple glob patterns can be used, i.e. `--exclude=foo-*.php` excludes any PHP file with the `foo-`
 *   prefix. Leading and trailing slashes are ignored, i.e. `/my/directory/` is the same as `my/directory`.
 *   The following files and folders are always excluded: node_modules, .git, .svn, .CVS, .hg, vendor, *.min.js.
 */
export interface Patterns {
	mergePaths?: string[]
	subtractPaths?: string[]
	subtractAndMerge?: boolean
	include: string[]
	exclude: string[]
}

/**
 * Create a POT file for a WordPress project.
 *
 * Scans PHP, Blade-PHP, and JavaScript files for translatable strings, as well as theme stylesheets and plugin files
 * if the source directory is detected as either a plugin or theme.
 *
 * @param {string} sourceDirectory - Directory to scan for string extraction.
 * @param {string} destination - Name of the resulting POT file.
 * @param {string | undefined} slug - Plugin or theme slug. Defaults to the source directory's basename.
 * @param {'plugin' | 'theme' | 'block' | 'theme-block' | 'generic'} domain - Text domain to look for in the source code,
 *   unless the `ignoreDomain` option is used. By default, the "Text Domain" header of the plugin or theme is used.
 *   If none is provided, it falls back to the project slug.
 * @param {boolean} ignoreDomain - Ignore the text domain completely and extract strings with any text domain.
 * @param {{}} headers - Array in JSON format of custom headers which will be added to the POT file. Defaults to empty array.
 * @param {boolean} location - Whether to write `#: filename:line` lines. Defaults to true, use `--no-location`
 *   to skip the removal. Note that disabling this option makes it harder for technically skilled translators
 *   to understand each messageâ€™s context.
 * @param {boolean} skipJs - Skips JavaScript string extraction. Useful when this is done in another build step, e.g. through Babel.
 * @param {boolean} skipPhp - Skips PHP string extraction.
 * @param {boolean} skipBlade - Skips Blade-PHP string extraction.
 * @param {boolean} skipBlockJson - Skips string extraction from block.json files.
 * @param {boolean} skipThemeJson - Skips string extraction from theme.json files.
 * @param {boolean} skipAudit - Skips string audit where it tries to find possible mistakes in translatable strings.
 *   Useful when running in an automated environment.
 * @param {string} fileComment - String that should be added as a comment to the top of the resulting POT file.
 *   By default, a copyright comment is added for WordPress plugins and themes.
 * @param {string} packageName - Name to use for the package name in the resulting POT file's `Project-Id-Version` header.
 *   Overrides the plugin or theme name, if applicable.
 * @param {boolean} silent - Whether to hide progress information.
 */
export interface Args {
	slug: string
	domain: DomainType
	paths: {
		cwd: string
		out: string
		root?: string
	}
	options: {
		ignoreDomain?: boolean
		silent?: boolean
		json?: boolean
		location?: boolean
		packageName?: string
		skip: {
			js?: boolean
			php?: boolean
			blade?: boolean
			blockJson?: boolean
			themeJson?: boolean
			audit?: boolean
		}
	}
	headers: Record<PotHeaders, string>
	patterns: Patterns
}

/**
 * Translation string metadata.
 * Gettext format: https://www.gnu.org/savannah-checkouts/gnu/gettext/FAQ.html
 *
 * @property {string} msgctxt - context for this translation, if not present the default context applies
 * @property {string} msgid - string to be translated
 * @property {string} msgid_plural the plural form of the original string (might not be present)
 * @property {string[]} msgstr  an array of translations
 * @property {{}} comments an object with the following properties: translator, reference, extracted, flag, previous.
 */
export interface TranslationStrings {
	[msgctxt: string]: { [msgId: string]: GetTextTranslation }
}

/** a json */
export type JsonData = Record<string, any>
