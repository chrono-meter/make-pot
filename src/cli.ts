#!/usr/bin/env node

import process from "node:process";
import * as yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { getArgs } from "./cli/getArgs.js";
import { getJsonArgs } from "./cli/getJsonArgs.js";
import makeJsonCommand from "./jsonCommand.js";
import makepotCommand from "./potCommand.js";
import type { Args, MakeJsonArgs } from "./types.js";

/** Main execution */
// Get the selected command
const r = yargs
	.default(hideBin(process.argv))
	.options({
		makejson: {
			describe: "Make JSON file",
			type: "boolean",
			default: false,
		},
	})
	.parseSync() as { makejson: boolean };

// Execute the command
if (!r.makejson) {
	makepotCommand(getArgs() as Args);
} else {
	makeJsonCommand(getJsonArgs() as MakeJsonArgs);
}
