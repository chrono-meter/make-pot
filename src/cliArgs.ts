import yargs from "yargs";
import {hideBin} from 'yargs/helpers'

export const args = yargs(hideBin(process.argv))
  .usage('Usage: $0 <source> [destination] [options]')
  .option('slug', {
    describe: 'Plugin or theme slug',
    type: 'string'
  })
  .option('domain', {
    describe: 'Text domain to look for in the source code',
    type: 'string'
  })
  .option('skip-js', {
    describe: 'Skip JavaScript files',
    type: 'boolean'
  })
  .option('skip-php', {
    describe: 'Skip PHP files',
    type: 'boolean'
  })
  .option('skip-blade', {
    describe: 'Skip Blade files',
    type: 'boolean'
  })
  .option('skip-block-json', {
    describe: 'Skip block.json files',
    type: 'boolean'
  })
  .option('skip-theme-json', {
    describe: 'Skip theme.json files',
    type: 'boolean'
  })
  .option('skip-audit', {
    describe: 'Skip auditing of strings',
    type: 'boolean'
  })
  .option('headers', {
    describe: 'Headers',
    type: 'string'
  })
  .option('file-comment', {
    describe: 'File comment',
    type: 'string'
  })
  .option('package-name', {
    describe: 'Package name',
    type: 'string'
  })
  .option('location', {
    describe: 'Include location information',
    type: 'boolean'
  })
  .option('ignore-domain', {
    describe: 'Ignore text domain',
    type: 'boolean'
  })
  .option('merge', {
    describe: 'Merge with existing POT file(s)',
    type: 'string'
  })
  .option('subtract', {
    describe: 'Subtract strings from existing POT file(s)',
    type: 'string'
  })
  .option('subtract-and-merge', {
    describe: 'Subtract and merge strings from existing POT file(s)',
    type: 'boolean'
  })
  .option('include', {
    describe: 'Include specific files',
    type: 'string'
  })
  .option('exclude', {
    describe: 'Exclude specific files',
    type: 'string'
  })
  .help('h')
  .alias('h', 'help')
  .argv
