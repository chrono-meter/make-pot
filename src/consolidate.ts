// flatten the results merging results into one array of strings
import type { TranslationStrings } from './types'

/**
 * Consolidates an array of translation strings into a single object.
 *
 * @returns The consolidated translation strings object.
 * @param translationsArray
 */

export function consolidate(
	translationsArray: TranslationStrings[]
): TranslationStrings {
	const mergedTranslations: TranslationStrings = {}

	translationsArray.forEach((translations) => {
		Object.entries(translations).forEach(
			([context, contextTranslations]) => {
				if (!mergedTranslations[context]) {
					mergedTranslations[context] = {}
				}

				Object.entries(contextTranslations).forEach(
					([msgid, translation]) => {
						if (!mergedTranslations[context][msgid]) {
							mergedTranslations[context][msgid] = {
								msgctxt: context !== '' ? context : undefined,
								msgid: msgid ?? '',
								msgid_plural: translation.msgid_plural,
								msgstr: translation.msgstr,
								comments: translation.comments,
							}
						}
					}
				)
			}
		)
	})

	return mergedTranslations
}
