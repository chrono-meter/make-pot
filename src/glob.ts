import Parser from 'tree-sitter'
import path from 'path'
import type { Args, Patterns } from './types'
import { Glob, Path } from 'glob'
import { minimatch } from 'minimatch'

// @ts-ignore
import * as Javascript from 'tree-sitter-javascript'
// @ts-ignore
import * as Ts from 'tree-sitter-typescript'
// @ts-ignore
import * as Php from 'tree-sitter-php'
import { detectPatternType } from './utils'

/**
 * Return the parser based on the file extension
 *
 * @param file - Path to the file
 * @return {Parser|{}|null} - the parser to be used with the file or null if no parser is found
 */
export function getParser(file: string): string | Parser {
	const ext = file.split('.').pop()
	switch (ext) {
		case 'ts':
			return Ts.default.typescript
		case 'tsx':
			return Ts.default.tsx
		case 'js':
		case 'jsx':
		case 'mjs':
		case 'cjs':
			return Javascript.default
		case 'php':
			return Php.default
		default:
			return ext!
	}
}

// Build the ignore function for Glob
export const ignoreFunc = (
	filePath: Path,
	excludedPatterns: string[]
): boolean => {
	return excludedPatterns.some((exclude) => {
		const type = detectPatternType(exclude)
		// return true to ignore
		switch (type) {
			case 'file':
				return filePath.isNamed(exclude)
			case 'directory':
				return filePath.relative().includes(exclude)
			default:
				// Handle glob patterns using minimatch
				return minimatch(filePath.relative(), exclude)
		}
	}) as boolean
}

/**
 * Retrieves a list of files based on the provided arguments and patterns.
 *
 * @param {Args} args - The argument object containing the source directory and other options.
 * @param {Patterns} pattern - The pattern object containing the included and excluded file patterns.
 * @return A promise that resolves to an array of file paths.
 */
export async function getFiles(args: Args, pattern: Patterns) {
	if (!args.options?.silent)
		console.log(
			'Searching in :',
			path.resolve(args.paths.cwd),
			'for ' + pattern.include.join(),
			'\nignoring patterns: ' + pattern.exclude.join()
		)

	// Execute the glob search with the built patterns
	return new Glob(pattern.include, {
		ignore: {
			ignored: (p: Path) => ignoreFunc(p, pattern.exclude),
		},
		nodir: true,
		cwd: args.paths.cwd,
		root: args.paths.root ? path.resolve(args.paths.root) : undefined,
	})
}
