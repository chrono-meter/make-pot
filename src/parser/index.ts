import path from "node:path";
import { extractMainFileData } from "../extractors/headers.js";
import { extractPackageJson } from "../extractors/json.js";
import { writeFile } from "../fs/index.js";
import type { Args } from "../types.js";
import { exec } from "./exec.js";

function getOutputPath(args: Args): string {
	return path.join(
		process.cwd(),
		args.headers?.domainPath ?? args.paths.out ?? "languages",
		`${args?.slug}.${args.options?.json ? "json" : "pot"}`,
	);
}

/**
 * Generates a pot file for localization.
 *
 * @param args - the command line arguments
 * @return {Promise<void>} - a promise that resolves when the pot file is generated
 */
export async function makePot(args: Args): Promise<string> {
	/** Collect metadata from the get package json */
	const pkgData = extractPackageJson(args);

	/** Get metadata from the main file (theme and plugin) */
	const metadata = extractMainFileData(args);

	/** Merge the metadata to get a single object with all the headers */
	args.headers = {
		...args.headers,
		...pkgData,
		...metadata,
	} as Args["headers"];

	/** Generate the pot file */
	const jsonTranslations = await exec(args);

	const outputPath = getOutputPath(args);

	console.log(`Writing pot file to ${outputPath}`);

	writeFile(jsonTranslations, outputPath);

	return jsonTranslations;
}
