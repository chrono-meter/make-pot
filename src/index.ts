#!/usr/bin/env node
import { makePot } from './makePot'

import { getArgs } from './cliArgs'

import yargs from 'yargs'

import { name, version } from '../package.json'

/** Main execution */
const args = getArgs()
if (Object.keys(args).length > 0) {
	console.log(name + ' version: ' + version)
	const timeStart = new Date()
	makePot(args)
		.then(() => {
			const timeEnd = new Date()
			console.log(
				`ðŸš€ Translation Pot file created in ${timeEnd.getTime() - timeStart.getTime()}ms`
			)
		})
		.catch((error) => {
			console.error(error)
		})
} else {
	yargs.showHelp()
}
