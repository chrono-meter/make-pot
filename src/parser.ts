import type { Args, Patterns, TranslationStrings } from './types'
import cliProgress, { type SingleBar } from 'cli-progress'
import { parseFile } from './extractors'
import { allowedFiles } from './const'
import { Glob, Path } from 'glob'
import { minimatch } from 'minimatch'
import Parser from 'tree-sitter'

// @ts-expect-error
import Js from 'tree-sitter-javascript'
// @ts-expect-error
import Php from 'tree-sitter-php'
// @ts-expect-error
import Ts from 'tree-sitter-typescript'
import path from 'path'
import gettextParser, { GetTextTranslations } from 'gettext-parser'
import { generateHeaderComments } from './utils'
import { consolidate } from './consolidate'

/**
 * Return the parser based on the file extension
 *
 * @param file - Path to the file
 * @return {Parser|{}|null} - the parser to be used with the file or null if no parser is found
 */
export function getParser(file: string): string | Parser {
	const ext = file.split('.').pop()
	switch (ext) {
		case 'ts':
			return Ts.typescript
		case 'tsx':
			return Ts.tsx
		case 'js':
		case 'jsx':
		case 'mjs':
		case 'cjs':
			return Js
		case 'php':
			return Php
		default:
			return ext!
	}
}

/**
 * Determines if a pattern represents a file, a directory, or a glob pattern.
 * @param pattern - The pattern string to evaluate.
 * @returns 'file', 'directory', or 'glob'.
 */
export function detectPatternType(
	pattern: string
): 'file' | 'directory' | 'glob' {
	const containsFileExtension = pattern.includes('.')
	const containsDirectorySeparator =
		pattern.includes(path.sep) || pattern.endsWith(path.sep)

	if (pattern.includes('*')) {
		return 'glob'
	} else if (!containsFileExtension && !containsDirectorySeparator) {
		return 'directory'
	} else if (containsFileExtension && !containsDirectorySeparator) {
		return 'file'
	} else {
		return 'glob'
	}
}

export function includeFunction(includePath: string[]) {
	return includePath.map((path) => {
		const type = detectPatternType(path)
		switch (type) {
			case 'directory':
				return '**/' + path + '/**'
			case 'file':
				return '**/' + path
			default:
				return path
		}
	})
}

/**
 * Retrieves a list of files based on the provided arguments and patterns.
 *
 * @param {Args} args - The argument object containing the source directory and other options.
 * @param {Patterns} pattern - The pattern object containing the included and excluded file patterns.
 * @return A promise that resolves to an array of file paths.
 */
export async function getFiles(args: Args, pattern: Patterns) {
	const includedPatterns = includeFunction(
		pattern.include.map((p) => p.trim()).filter((p) => p)
	) ?? ['**']

	// Process excludePaths
	const excludedPatterns = pattern.exclude ?? []

	// Build the ignore function for Glob
	const ignoreFunc = (filePath: Path): boolean => {
		return excludedPatterns.some((exclude) => {
			const type = detectPatternType(exclude)
			switch (type) {
				case 'file':
					return filePath.name === exclude
				case 'directory':
					return filePath.path.includes(exclude)
				default:
					// Handle glob patterns using minimatch or a similar library
					return minimatch(filePath.path, exclude)
			}
		}) as boolean
	}

	if (!args.silent)
		console.log(
			'Searching in :',
			path.resolve(args.sourceDirectory),
			'for ' + includeFunction(includedPatterns).join(),
			'\nExcluding : ' + excludedPatterns.join()
		)

	// Execute the glob search with the built patterns
	return new Glob(includedPatterns, {
		ignore: {
			ignored: (p) => {
				return ignoreFunc(p)
			},
		},
		nodir: true,
		cwd: args.sourceDirectory,
	})
}

/**
 * Initializes a progress bar and returns the progress bar element.
 *
 * @param {Args} args - The argument object containing the source directory and other options.
 * @param {number} filesCount - An array of file names.
 * @return {cliProgress.SingleBar} The progress bar element.
 */
function initProgress(args: Args, filesCount: number): SingleBar | null {
	if (args.silent) return null
	// Set up the progress bar
	const progressBar = new cliProgress.SingleBar(
		{
			clearOnComplete: true,
			etaBuffer: 1000,
			hideCursor: true,
			format: ' {bar} {percentage}% | ETA: {eta}s | {filename} | {value}/{total}',
		},
		cliProgress.Presets.shades_classic
	)

	progressBar.start(filesCount, 0)

	// Return the progress bar element
	return progressBar
}

/**
 * Retrieves an array of translation strings from files that match the specified arguments and pattern.
 *
 * @param {Args} args - The arguments to specify which files to search and parse.
 * @param {string[]} files - An array of file names to search and parse.
 * @return {Promise<TranslationStrings[]>} A promise that resolves to an array of translation strings found in the files.
 */
export async function getStrings(
	args: Args,
	files: Glob<{}>
): Promise<TranslationStrings> {
	const tasks: Promise<TranslationStrings>[] = []

	const progressBar = initProgress(
		args,
		Array.from(files.iterateSync()).length
	)

	// loop through the files and parse them
	for (const file of files) {
		// get the file extension
		const ext = file.split('.').pop() || 'undefined'
		// check if the extension is allowed
		if (!allowedFiles.includes(ext)) {
			// log the filepath
			if (progressBar) {
				progressBar.increment(1, {
					filename: `Skipping ${ext} (not a valid file extension)`,
				})
			}
			continue
		}

		const task = parseFile({
			filepath: path.join(process.cwd(), args.sourceDirectory, file),
			language: getParser(file),
		}) as Promise<TranslationStrings>

		// log the filepath
		if (progressBar) {
			progressBar.increment(1, {
				filename: file,
			})
		}

		// add the task to the array if it's not null
		if (task !== null) tasks.push(task as Promise<TranslationStrings>)
	}

	const results = await Promise.all(tasks)

	// stop the progress bar if it's not silent
	if (progressBar) progressBar.stop()

	const mergedResult = consolidate(results.filter((r) => r !== null).flat())

	console.log('Strings grouped')

	if (!args.silent) {
		console.log(
			'📝 Found',
			Object.values(mergedResult).length,
			'group of strings in',
			results.length,
			'files.\n',
			'In total ' +
				Object.values(mergedResult)
					.map((v) => Object.keys(v).length)
					.reduce((acc, val) => acc + val, 0) +
				' strings were found'
		)
	}

	return mergedResult
}

/**
 * Runs the extract process based on the given arguments.
 *
 * @param {Args} args - The arguments for the extract process.
 * @return {Promise<string>} - A promise that resolves with the extracted data.
 */
export async function runExtract(args: Args) {
	const pattern = {
		include: args.include || [],
		exclude: args.exclude || [],
		mergePaths: args.mergePaths,
		subtractPaths: args.subtractPaths,
		subtractAndMerge: args.subtractAndMerge,
	} as Patterns

	// Additional logic to handle different file types and formats
	// Exclude blade.php files if --skip-blade is set
	if (args.skipPhp !== undefined || args.skipBlade !== undefined) {
		if (args.skipBlade !== undefined) {
			// php files but not blade.php
			pattern.include.push('**/*.php')
		} else {
			pattern.include.push('**/*.php', '!**/blade.php')
		}
	}

	// js typescript mjs cjs etc
	if (args.skipJs !== undefined) {
		pattern.include.push('**/*.{js,jsx,ts,tsx,mjs,cjs}')
	}

	if (args.skipBlockJson !== undefined) {
		pattern.include.push('block.json')
	}

	if (args.skipThemeJson !== undefined) {
		pattern.include.push('theme.json')
	}

	const files = await getFiles(args, pattern)
	const stringsJson = await getStrings(args, files)

	const additionalHeaders = {
		'content-type': 'text/plain; charset=iso-8859-1',
		...args.headers,
		'plural-forms': 'nplurals=2; plural=(n != 1);',
	}

	const buffer = gettextParser.po.compile(
		{
			headers: additionalHeaders,
			comments: generateHeaderComments(args),
			translations: stringsJson,
			charset: 'iso-8859-1',
		} as unknown as GetTextTranslations,
		{
			sort: true,
		}
	)

	return buffer.toString('utf-8') as string
}
