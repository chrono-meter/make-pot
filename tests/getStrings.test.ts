import { describe, expect } from '@jest/globals'
import { getStrings } from '../src/parser'
import { Args, DomainType } from '../src/types'
import { getFiles } from '../lib/glob'

const args: Args = {
	slug: 'plugin-slug',
	paths: { cwd: 'tests/fixtures/sourcedir/', out: 'tests/fixtures/' },
	domain: 'plugin' as DomainType,
	patterns: {
		include: ['**'],
		exclude: [
			'node_modules',
			'vendor',
			'dist',
			'tests',
			'package-lock.json',
		],
	},
}

describe('getStrings', () => {
	it('Should build pot file', async () => {
		const files = await getFiles(args as Args, {
			include: ['file.php'],
			exclude: ['node_modules', 'dist'],
		})
		const dataExtracted = await getStrings({ ...args } as Args, files)
		const expected = {
			'': {
				sdasdasdasd: {
					msgid: 'sdasdasdasd',
					msgid_plural: undefined,
					msgstr: [],
				},
				'strong magenta': {
					msgid: 'strong magenta',
					msgid_plural: undefined,
					msgstr: [],
				},
				'light grayish magenta': {
					msgid: 'light grayish magenta',
					msgid_plural: undefined,
					msgstr: [],
				},
				'very light gray': {
					msgid: 'very light gray',
					msgid_plural: undefined,
					msgstr: [],
				},
				'very dark gray': {
					msgid: 'very dark gray',
					msgid_plural: undefined,
					msgstr: [],
				},
				'Vivid cyan blue to vivid purple': {
					msgid: 'Vivid cyan blue to vivid purple',
					msgid_plural: undefined,
					msgstr: [],
				},
				'Vivid green cyan to vivid cyan blue': {
					msgid: 'Vivid green cyan to vivid cyan blue',
					msgid_plural: undefined,
					msgstr: [],
				},
				'Light green cyan to vivid green cyan': {
					msgid: 'Light green cyan to vivid green cyan',
					msgid_plural: undefined,
					msgstr: [],
				},
				'Luminous vivid amber to luminous vivid orange': {
					msgid: 'Luminous vivid amber to luminous vivid orange',
					msgid_plural: undefined,
					msgstr: [],
				},
				'Luminous vivid orange to vivid red': {
					msgid: 'Luminous vivid orange to vivid red',
					msgid_plural: undefined,
					msgstr: [],
				},
				Small: {
					msgid: 'Small',
					msgid_plural: undefined,
					msgstr: [],
				},
				Regular: {
					msgid: 'Regular',
					msgid_plural: undefined,
					msgstr: [],
				},
				Large: {
					msgid: 'Large',
					msgid_plural: undefined,
					msgstr: [],
				},
				Huge: {
					msgid: 'Huge',
					msgid_plural: undefined,
					msgstr: [],
				},
			},
		}
		expect(dataExtracted).toMatchObject(expected)
	})
	it('Should build pot file from fixtures block.json', async () => {
		const currentArgs = {
			...args,
			paths: { cwd: 'tests/fixtures/block/', out: 'tests/fixtures/' },
			patterns: {
				include: ['block.json'],
				exclude: ['node_modules'],
			},
		} as Args
		const files = await getFiles(currentArgs as Args, currentArgs.patterns)
		const dataExtracted = await getStrings(currentArgs as Args, files)

		const expected = {
			'block variation keyword': {
				undefined: {
					msgstr: [],
					msgid: 'undefined',
					msgctxt: 'block variation keyword',
				},
			},
			'block variation title': {
				title: {
					msgstr: [],
					msgid: 'title',
					msgctxt: 'block variation title',
				},
			},
			'block variation description': {
				description: {
					msgstr: [],
					msgid: 'description',
					msgctxt: 'block variation description',
				},
			},
			'block style label': {
				label: {
					msgstr: [],
					msgid: 'label',
					msgctxt: 'block style label',
				},
			},
		}
		expect(dataExtracted).toMatchObject(expected)
	})
})
