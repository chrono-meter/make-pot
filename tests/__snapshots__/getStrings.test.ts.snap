// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStrings Should build pot file 1`] = `
{
  "": {
    "Huge": {
      "comments": {
        "reference": "file.php:252",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Huge",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Large": {
      "comments": {
        "reference": "file.php:247",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Large",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Light green cyan to vivid green cyan": {
      "comments": {
        "reference": "file.php:189",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Light green cyan to vivid green cyan",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid amber to luminous vivid orange": {
      "comments": {
        "reference": "file.php:194",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Luminous vivid amber to luminous vivid orange",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid orange to vivid red": {
      "comments": {
        "reference": "file.php:199",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Luminous vivid orange to vivid red",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Regular": {
      "comments": {
        "reference": "file.php:242",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Regular",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Small": {
      "comments": {
        "reference": "file.php:237",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Small",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid cyan blue to vivid purple": {
      "comments": {
        "reference": "file.php:179",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Vivid cyan blue to vivid purple",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid green cyan to vivid cyan blue": {
      "comments": {
        "reference": "file.php:184",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Vivid green cyan to vivid cyan blue",
      "msgid_plural": "",
      "msgstr": [],
    },
    "light grayish magenta": {
      "comments": {
        "reference": "file.php:123",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "light grayish magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "sdasdasdasd": {
      "comments": {
        "reference": "file.php:7",
        "translator": "ciao2!",
      },
      "msgctxt": undefined,
      "msgid": "sdasdasdasd",
      "msgid_plural": "",
      "msgstr": [],
    },
    "strong magenta": {
      "comments": {
        "reference": "file.php:118",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "strong magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very dark gray": {
      "comments": {
        "reference": "file.php:133",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "very dark gray",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very light gray": {
      "comments": {
        "reference": "file.php:128",
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "very light gray",
      "msgid_plural": "",
      "msgstr": [],
    },
  },
}
`;
